pyodide-build/pyodide_build/buildall.py

```diff
@@ -158,15 +164,95 @@ def build(self, args: Any) -> None:
             )
 
         if p.returncode != 0:
-            print(f"Error building {self.name}. Printing build logs.")
+            logger.error(f"Error building {self.name}. Printing build logs.")
 
             with open(self.pkgdir / "build.log") as f:
-                shutil.copyfileobj(f, sys.stdout)
+                shutil.copyfileobj(f, sys.stderr)
 
-            print("ERROR: cancelling buildall")
+            logger.error("ERROR: cancelling buildall")
             raise BuildError(p.returncode)
 
 
+class PackageStatus:
+    def __init__(
+        self, *, name: str, idx: int, thread: int, total_packages: int
+    ) -> None:
+        self.pkg_name = name
+        self.prefix = f"[{idx}/{total_packages}] " f"(thread {thread})"
+        self.status = Spinner("dots", style="red", speed=0.2)
+        self.table = Table.grid(padding=1)
+        self.table.add_row(f"{self.prefix} building {self.pkg_name}", self.status)
+        self.finished = False
+
+    def finish(self, success: bool, elapsed_time: float) -> None:
+        time = datetime.utcfromtimestamp(elapsed_time)
+        if time.minute == 0:
+            minutes = ""
+        else:
+            minutes = f"{time.minute}m "
+        timestr = f"{minutes}{time.second}s"
+
+        status = "built" if success else "failed"
+        done_message = f"{self.prefix} {status} {self.pkg_name} in {timestr}"
+
+        self.finished = True
+
+        if success:
+            logger.success(done_message)
+        else:
+            logger.error(done_message)
+
+    def __rich__(self):
+        return self.table
+
+
+class ReplProgressFormatter:
+    def __init__(self, num_packages: int) -> None:
+        self.progress = Progress(
+            "[progress.description]{task.description}",
+            BarColumn(),
+            "{task.completed}/{task.total} [progress.percentage]{task.percentage:>3.0f}%",
+            "Time elapsed:",
+            TimeElapsedColumn(),
+        )
+        self.task = self.progress.add_task("Building packages...", total=num_packages)
+        self.packages: list[PackageStatus] = []
+        self.reset_grid()
+
+    def reset_grid(self):
+        """Empty out the rendered grids."""
+        self.top_grid = Table.grid()
+
+        for package in self.packages:
+            if not package.finished:
```

Comment:
Since we call `remove_package` immediately after `finish`, this conditional probably isn't needed?